src/Feldspar/Onnx/Operators.hs

```diff
@@ -244,23 +244,23 @@ bcDivI :: (Integral a, ShapelyU sh1 sh2)
 bcDivI = bcZipWith div -- Or quot???
 
 -- | Implementation of ONNX batch normalization
-onnxBatchNormalization :: Floating a
-                       => Attrs  -- ^ attributes (including epsilon)
-                       -> DPull DIM4 a -- ^ data
-                       -> DPull DIM1 a -- ^ gamma
-                       -> DPull DIM1 a -- ^ beta
-                       -> DPull DIM1 a -- ^ mean
-                       -> DPull DIM1 a -- ^ var
-                       -> DPull DIM4 a
-onnxBatchNormalization attrs xs gamma beta mean var = ys
+onnxBatchNormalization_5 :: Floating a
+                         => Attrs  -- ^ attributes (including epsilon)
+                         -> DPull DIM4 a -- ^ data
+                         -> DPull DIM1 a -- ^ gamma
+                         -> DPull DIM1 a -- ^ beta
+                         -> DPull DIM1 a -- ^ mean
+                         -> DPull DIM1 a -- ^ var
+                         -> DPull DIM4 a
+onnxBatchNormalization_5 attrs xs gamma beta mean var = ys
   where invDev = map (\ v -> 1.0 / sqrt (v + epsilon)) var <! 1 <! 1
         xsHat = bcMul invDev $ bcSub xs $ mean <! 1 <! 1
         ys = bcAdd (bcMul xsHat $ gamma <! 1 <! 1) $ beta <! 1 <! 1
         epsilon = value $ P.realToFrac $ getAttr attrs aaFloat 1e-5 "epsilon"
 
 -- | Flatten a tensor to a matrix
-onnxFlatten :: Pushy vec => Attrs -> vec a -> Push DIM2 a
-onnxFlatten attrs vec = flatPush (P.fromIntegral $ getAttr attrs aaInt 1 "axis") $ toPush vec
+onnxFlatten_1 :: (Pushy vec, Syntax a) => Attrs -> vec a -> Pull DIM2 a
```

Comment:
Everything seems to be Pully in this module except this stuff. Why do we need to store a reshape (index transformation) to memory?

Reply:
This could in principle be done with a Pull vector and consequently be fuseable. The problem is that the index expressions would be horrible, using division and modulus since the result has fewer dimensions than the argument so that a single integer index must be split into several. We have no optimization in place that eliminates them, so they will typically end up in the innermost loop.

Reply:
Would be great to have that kind of information in a comment in the source code.